dev/NumberBox/APITests/NumberBoxTests.cs

```diff
@@ -0,0 +1,99 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using Common;
+using MUXControlsTestApp.Utilities;
+using Windows.UI.Xaml.Controls;
+using Windows.UI.Xaml.Media;
+using System.Linq;
+using System.Collections.Generic;
+using MUXControlsTestApp;
+using Microsoft.UI.Xaml.Controls;
+using Windows.UI.Xaml.Controls.Primitives;
+
+#if USING_TAEF
+using WEX.TestExecution;
+using WEX.TestExecution.Markup;
+using WEX.Logging.Interop;
+#else
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+using Microsoft.VisualStudio.TestTools.UnitTesting.Logging;
+#endif
+
+namespace Windows.UI.Xaml.Tests.MUXControls.ApiTests
+{
+    [TestClass]
+    public class NumberBoxTests : ApiTestBase
```

Comment:
>ApiTestBase [](start = 34, length = 11)

ApiTestBase gives you access to the Content Property and the Content.UpdateLayout() method. Using these will significantly reduce the complexity of this test. You can see examples in a lot (most?) of our api tests, specifically here is one: [https://github.com/microsoft/microsoft-ui-xaml/blob/master/dev/Repeater/APITests/LayoutTests.cs](https://github.com/microsoft/microsoft-ui-xaml/blob/master/dev/Repeater/APITests/LayoutTests.cs)

Reply:
Using this IdleSynchronizer.Wait() and exiting the UIThread shouldn't be needed

---
In reply to: [405702960](https://github.com/microsoft/microsoft-ui-xaml/pull/2224#discussion_r405702960) [](ancestors = 405702960)

Reply:
@Felix-Dev this is the last comment I have, then I'll sign off :)

Reply:
@StephenLPeters There seems to be a timing issue here using this approach in one specific test (code slightly abbreviated to only show relevant part):
```cs
[TestMethod]
public void VerifyNumberBoxCornerRadius2()
{
    var numberBox = SetupNumberBox2();

    RepeatButton spinButtonDown = null;
    TextBox textBox = null;
    RunOnUIThread.Execute(() =>
    {
        // first test: Uniform corner radius of '2' with no spin buttons shown
        numberBox.SpinButtonPlacementMode = NumberBoxSpinButtonPlacementMode.Hidden;
        numberBox.CornerRadius = new CornerRadius(2);

        Content.UpdateLayout();

        textBox = TestUtilities.FindDescendents<TextBox>(numberBox).Where(e => e.Name == "InputBox").Single();
        Verify.AreEqual(new CornerRadius(2, 2, 2, 2), textBox.CornerRadius);

        // second test: Uniform corner radius of '2' with spin buttons in inline mode (T-rule applies now)
        numberBox.SpinButtonPlacementMode = NumberBoxSpinButtonPlacementMode.Inline;

        Content.UpdateLayout();

        spinButtonDown = TestUtilities.FindDescendents<RepeatButton>(numberBox).Where(e => e.Name == "DownSpinButton").Single();

        Verify.AreEqual(new CornerRadius(2, 0, 0, 2), textBox.CornerRadius);
        Verify.AreEqual(new CornerRadius(0, 2, 2, 0), spinButtonDown.CornerRadius);

        // third test: Set uniform corner radius of '4' with spin buttons in inline mode
        // This check makes sure that updating the CornerRadius values of the numberbox in inline mode
        // does not break the T-rule.
        numberBox.CornerRadius = new CornerRadius(4);

        Content.UpdateLayout();

        // TEST FAIL HERE: actual corner radius is <4,4,4,4> so T-rule broken
        // Seems like the visual state we are entering when corner radius changes in Inline mode
        // is not applied yet
        Verify.AreEqual(new CornerRadius(4, 0, 0, 4), textBox.CornerRadius);
        Verify.AreEqual(new CornerRadius(0, 4, 4, 0), spinButtonDown.CornerRadius);
    });
}
```

See the comment for the final test above (setting CornerRadius to 4 in the third test).

The test passes as expected when I write it like this:
```cs
[TestMethod]
public void VerifyNumberBoxCornerRadius3()
{
    var numberBox = SetupNumberBox2();

    RepeatButton spinButtonDown = null;
    TextBox textBox = null;
    RunOnUIThread.Execute(() =>
    {
        // first test: Uniform corner radius of '2' with no spin buttons shown
        numberBox.SpinButtonPlacementMode = NumberBoxSpinButtonPlacementMode.Hidden;
        numberBox.CornerRadius = new CornerRadius(2);

        Content.UpdateLayout();

        textBox = TestUtilities.FindDescendents<TextBox>(numberBox).Where(e => e.Name == "InputBox").Single();
        Verify.AreEqual(new CornerRadius(2, 2, 2, 2), textBox.CornerRadius);

        // second test: Uniform corner radius of '2' with spin buttons in inline mode (T-rule applies now)
        numberBox.SpinButtonPlacementMode = NumberBoxSpinButtonPlacementMode.Inline;

        Content.UpdateLayout();

        spinButtonDown = TestUtilities.FindDescendents<RepeatButton>(numberBox).Where(e => e.Name == "DownSpinButton").Single();

        Verify.AreEqual(new CornerRadius(2, 0, 0, 2), textBox.CornerRadius);
        Verify.AreEqual(new CornerRadius(0, 2, 2, 0), spinButtonDown.CornerRadius);

        // third test: Set uniform corner radius of '4' with spin buttons in inline mode
        numberBox.CornerRadius = new CornerRadius(4);
    });
    IdleSynchronizer.Wait();

    RunOnUIThread.Execute(() =>
    {
         // This check makes sure that updating the CornerRadius values of the numberbox in inline mode
        // does not break the T-rule.
        Verify.AreEqual(new CornerRadius(4, 0, 0, 4), textBox.CornerRadius);
        Verify.AreEqual(new CornerRadius(0, 4, 4, 0), spinButtonDown.CornerRadius);
    });
}
```

Reply:
Strange, @ranjeshj I wouldn't expect a behavioral difference here.

Reply:
@Felix-Dev I still prefer the updated, working code above.

Reply:
@StephenLPeters Updated the test. Please take a look. (I added one more check in the API test.)